File: 0103-binary-tree-zigzag-level-order-traversal/0103-binary-tree-zigzag-level-order-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:

    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        vector<vector<int>> v;
        if(root==NULL)
        return v;

        queue<TreeNode*> q;
        q.push(root);
        bool zz=true;

        while(!q.empty())
        {
            vector<int> ans;
            int size=q.size();
            for(int i=0;i<size;i++)
            {
                TreeNode *temp=q.front();
                q.pop();
                ans.push_back(temp->val);

                if(temp->left)
                q.push(temp->left);

                if(temp->right)
                q.push(temp->right);       
            }

            if(zz)
            {
                v.push_back(ans);
                zz=!zz;
            }
            else
            {
                reverse(ans.begin(),ans.end());
                v.push_back(ans);
                zz=!zz;
            }

        }
        return v;
    }
};

File: 0103-binary-tree-zigzag-level-order-traversal/0103-binary-tree-zigzag-level-order-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        List<List<Integer>> res = new LinkedList<>();
        if(root == null) return res;
        
        Queue<TreeNode> q = new LinkedList<>();
        q.add(root);
        
        int level = 0;
        
        while(!q.isEmpty()){
            int size = q.size();
            List<Integer> levelSubList = new LinkedList<>();
            while(size > 0){
                TreeNode node = q.poll();
                if(level%2 != 0){
                    levelSubList.addFirst(node.val);
                }
                else{
                    levelSubList.add(node.val);
                }
                
                if(node.left != null) q.add(node.left);
                if(node.right != null) q.add(node.right);
                
                size--;
            }
            level++;
            res.add(levelSubList);
        }
        return res;
    }
}